admin/test/test_merge_pr.py

```diff
@@ -0,0 +1,212 @@
+        assume(status['state'] != u"success")
+        self.assertEqual(True, merge_pr.not_success(status))
+
+    @given(commit_statuses)
+    def test_not_success_success(self, status):
+        status['state'] = u"success"
+        self.assertEqual(False, merge_pr.not_success(status))
+
+    @given(commit_statuses, jenkins_results)
+    def test_format_status(self, status, jenkins):
+        formatted = merge_pr.format_status((status, jenkins))
+        self.assertIsInstance(formatted, unicode)
+        self.assertIn(status['context'], formatted)
+        self.assertIn(status['target_url'], formatted)
+
+
+# These strategies are far from complete coverage of the possible
+# Jenkins API responses.
+
+jenkins_builds = fixed_dictionaries(dict(
+    result=sampled_from([None, 'FAILURE', 'ABORTED', 'SUCCESS'])))
+"""Strategy for generating records of individual builds of a Jenkins job."""
+
+
+jenkins_build_result_with_builds = fixed_dictionaries(dict(
+    inQueue=booleans(),
+    builds=lists(jenkins_builds),
+    property=dictionaries(text(), text())))
+"""Strategy for generating Jenkins API information for a job with builds."""
+
+jenkins_build_result_with_no_builds = fixed_dictionaries(dict(
+    inQueue=booleans()))
+"""Strategy for generating Jenkins API information for a job with no builds."""
+
+jenkins_build_results = one_of(
+    jenkins_build_result_with_builds,
+    jenkins_build_result_with_no_builds,
+    just(pmap()))
+"""Strategy for generating Jenkins API information for a job."""
+
+
+class JenkinsResultsTests(TestCase):
+    """
+    Tests for interpretation of build results from Jenkins.
+    """
+
+    @given(jenkins_build_results)
+    def test_result_types(self, info):
+        result, params = merge_pr.jenkins_info_from_response(info)
+        self.assertIn(result, list(merge_pr.JenkinsResults.iterconstants()))
+        if params is not None:
+            self.assertIsInstance(params, dict)
+
+    @given(jenkins_build_results)
+    def test_in_queue(self, info):
+        info = dict(info)
+        info['inQueue'] = True
+        result, params = merge_pr.jenkins_info_from_response(info)
+        self.assertEqual(merge_pr.JenkinsResults.RUNNING, result)
+        self.assertEqual({}, params)
+
+    @given(jenkins_build_results)
+    def test_builds_not_present(self, info):
+        assume(not info.get('inQueue', False))
+        info = dict(info)
+        if 'builds' in info:
+            del info['builds']
+        result, params = merge_pr.jenkins_info_from_response(info)
+        self.assertEqual(merge_pr.JenkinsResults.UNKNOWN, result)
+        self.assertEqual({}, params)
+
+    @given(jenkins_build_results)
+    def test_no_builds(self, info):
+        assume(not info.get('inQueue', False))
+        info = dict(info)
+        info['builds'] = []
+        result, params = merge_pr.jenkins_info_from_response(info)
+        self.assertEqual(merge_pr.JenkinsResults.NOTRUN, result)
+        self.assertEqual({}, params)
```

Comment:
Hrm... I actually struggle with this one as far as deciding if hypothesis adds test coverage or just complexity over single example tests. It seems after `assume`s and mutations, hypothesis is usually left to fiddle with keys on the dictionaries that obviously have not bearing on the output.

It's tempting to have just a single test that asserts different things depending on the example hypothesis generates, but then you are just re-implementing your production code in your test... And you would still sort of want the single example style test to verify the logic in your test code.

So there are 2 concerns with using hypothesis in my mind, 1 is code complexity and the other is test runtime. Obviously these have to be balanced against the enhanced code coverage. You have some nice clean strategies here, so I think the complexity is well within our budget, but I'm a little concerned about test time (especially with use of assume and the mutations).

I noticed running this class of test suites takes significantly longer than the rest of the file. 4 seconds isn't great for testing pure-functional code. Perhaps consider trying to get that down? Simplifying the strategy for the property dict I think has the best bang for your buck, something like `property=dictionaries(text(max_size=2), text(max_size=2), max_size=2)` runs a lot faster, and I don't think it cheapens your test that much if at all.

```
flocker-review ♦ ~/dev/flocker-review (merge-pr-no-builds-FLOC-4140)$ trial admin.test.test_merge_pr.JenkinsResultsTests                                                                                    
admin.test.test_merge_pr
  JenkinsResultsTests
    test_builds_not_present ...                                            [OK]
    test_in_queue ...                                                      [OK]
    test_no_builds ...                                                     [OK]
    test_result_types ...                                                  [OK]

-------------------------------------------------------------------------------
Ran 4 tests in 3.838s

PASSED (successes=4)
flocker-review ♦ ~/dev/flocker-review (merge-pr-no-builds-FLOC-4140)$ trial admin.test.test_merge_pr.{SmokeTests,StatusesTests,URLTests}
admin.test.test_merge_pr
  SmokeTests
    test_help ...                                                          [OK]
  StatusesTests
    test_final_status_many ...                                             [OK]
    test_final_status_one ...                                              [OK]
    test_format_status ...                                                 [OK]
    test_not_success ...                                                   [OK]
    test_not_success_success ...                                           [OK]
  URLTests
    test_pr_api_url ...                                                    [OK]
    test_url_path ...                                                      [OK]
    test_url_path_no_hostname ...                                          [OK]
    test_url_path_parts ...                                                [OK]
    test_url_path_parts_no_hostname ...                                    [OK]

-------------------------------------------------------------------------------
Ran 11 tests in 0.610s

PASSED (successes=11)
```